File: 72-edit-distance/edit-distance.cpp
class Solution {
public:
    int dp[501][501];
    long fun(string word1, string word2,int i,int j)
    {
        if(i>=word1.size() && j>=word2.size())
        {
            return 0;
        }
        if(i>=word1.size() ||  j>=word2.size())
        {
            return max(word2.size() - j,word1.size() - i);
        }
        if(dp[i][j]!=-1)
        {
            return dp[i][j];
        }
        long ans=0;
        if(word1[i]==word2[j])
        {
            ans=fun(word1,word2,i+1,j+1);
        }
       else
        {
            ans=min({1+fun(word1,word2,i+1,j),1+fun(word1,word2,i+1,j+1),1+fun(word1,word2,i,j+1)});
        }
        return dp[i][j]=ans;
    }
    int minDistance(string word1, string word2) {
        memset(dp,-1,sizeof(dp));
        return fun(word1,word2,0,0);
    }
};

File: 72-edit-distance/edit-distance.java
class Solution {
    private int lcs(char[] a,char[] b,int i,int j,int[][] dp)
    {
        if(i==a.length)
        {
            return b.length-j;
        }
        if(j==b.length)
        {
            return a.length - i;
        }
        if(dp[i][j]!=-1)
        {
            return dp[i][j];
        }
        int x = 0,y=0;
        if(a[i]==b[j])
        {
            return lcs(a,b,i+1,j+1,dp);
        }
        int insert = lcs(a,b,i,j+1,dp);
        int delete = lcs(a,b,i+1,j,dp);
        int replace = lcs(a,b,i+1,j+1,dp);
        return dp[i][j]=1+Math.min(insert,Math.min(delete,replace));

    }
    public int minDistance(String word1, String word2) {
        char[] a = word1.toCharArray();
        char[] b = word2.toCharArray();
        int n=a.length;
        int m=b.length;
        int[][] dp = new int[n][m];
        for(int[] i : dp)
        {
            Arrays.fill(i,-1);
        }
        int x=lcs(a,b,0,0,dp);

        return x;
    }
}